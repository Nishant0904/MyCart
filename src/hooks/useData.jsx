import { useEffect, useState } from 'react'

const useData = (path) => {
    const [data, setData] = useState(null)
    const [error, setError] = useState(null)
    const [loading, setLoading] = useState(true)

    useEffect(() => {
        fetch(`https://fakestoreapi.com${path}`)
            .then((response) => {
                if (response.status >= 400) {
                    throw new Error('server error')
                }
                return response.json()
            })
            .then((response) => setData(response))
            .catch((error) => setError(error))
            .finally(() => setLoading(false))
    }, [path])

    return { data, error, loading }
}

export default useData
