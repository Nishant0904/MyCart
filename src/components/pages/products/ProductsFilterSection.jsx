import styled from 'styled-components'
import SkeletonLoadingText from '../loading/SkeletonLoadingText'
import useData from '../../../hooks/useData'
import capitalize from '../../../utils/capitalize'
import PropTypes from 'prop-types'

const ProductsFilterSection = ({
    pageTitle,
    handleSearchInputChange,
    handleCategoryChange,
    handleSort,
}) => {
    const { data, error, loading } = useData('/products/categories')

    if (error)
        return (
            <StyledProductsFilterSection>
                <p>An error occurred. Please check back later.</p>
            </StyledProductsFilterSection>
        )
    if (loading)
        return (
            <StyledProductsFilterSection>
                <SkeletonLoadingText data-testid='skeleton-loading-text' />
                <ProductsFiltersContainer>
                    <SkeletonLoadingText />
                    <SkeletonLoadingText />
                    <SkeletonLoadingText />
                </ProductsFiltersContainer>
            </StyledProductsFilterSection>
        )

    return (
        <StyledProductsFilterSection data-testid='products-filter-section'>
            <h1>{pageTitle}</h1>
            <ProductsFiltersContainer>
                <div>
                    <label htmlFor='search-input'>Search:</label>
                    <input
                        type='text'
                        autoComplete='off'
                        id='search-input'
                        onChange={handleSearchInputChange}
                    />
                </div>
                <div>
                    <label htmlFor='category-select'>Category:</label>
                    <select
                        id='category-select'
                        onChange={handleCategoryChange}
                    >
                        <option value='all'>All</option>
                        {data.map((category, index) => {
                            return (
                                <option key={index} value={category}>
                                    {capitalize(category)}
                                </option>
                            )
                        })}
                    </select>
                </div>
                <div>
                    <label htmlFor='sort-select'>Sort:</label>
                    <select id='sort-select' onChange={handleSort}>
                        <option value='id-ascending'>Default (by SKU)</option>
                        <option value='title-ascending'>A-Z</option>
                        <option value='title-descending'>Z-A</option>
                        <option value='price-ascending'>
                            Price (low to high)
                        </option>
                        <option value='price-descending'>
                            Price (high to low)
                        </option>
                        <option value='rating-descending'>
                            Rating (high to low)
                        </option>
                    </select>
                </div>
            </ProductsFiltersContainer>
        </StyledProductsFilterSection>
    )
}

const StyledProductsFilterSection = styled.section`
    padding: 48px 96px;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    gap: 24px;
    background-color: var(--secondary-bg-color);
    border-bottom: 1px solid black;

    @media (max-width: 600px) {
        padding: 12px 16px;
    }
`

const ProductsFiltersContainer = styled.div`
    display: flex;
    align-items: center;
    gap: 16px;

    div {
        display: flex;
        align-items: center;
        gap: 4px;
    }

    label {
        font-size: 0.8rem;
        font-weight: 900;
        text-transform: uppercase;
        letter-spacing: 0.1rem;
    }

    input,
    select {
        border: 1px solid grey;
        padding: 4px;
    }

    @media (max-width: 740px) {
        flex-direction: column;
        align-items: start;
        width: 100%;

        div {
            width: 100%;
        }

        input, select {
            width: 100%;
        }
    }
`

ProductsFilterSection.propTypes = {
    pageTitle: PropTypes.string,
    handleCategoryChange: PropTypes.func,
    handleSearchInputChange: PropTypes.func,
    handleSort: PropTypes.func,
}

export default ProductsFilterSection
