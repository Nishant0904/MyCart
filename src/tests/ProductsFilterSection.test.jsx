import { describe, it, expect, vi } from 'vitest'
import { act } from 'react-dom/test-utils'
import { render, screen } from '@testing-library/react'
import useData from '../hooks/useData'
import ProductsFilterSection from '../components/pages/products/ProductsFilterSection'
import userEvent from '@testing-library/user-event'

vi.mock('../hooks/useData', () => ({
    __esModule: true,
    default: vi.fn(),
}))

const mockCategories = [
    'electronics',
    'jewelery',
    "men's clothing",
    "women's clothing",
]

describe('Products Filter Section test', () => {
    it('renders error message when data is null', async () => {
        useData.mockImplementation(() => {
            return { data: null, error: true, loading: false }
        })

        act(() => {
            render(<ProductsFilterSection />)
        })

        expect(
            screen.getByText('An error occurred. Please check back later.'),
        ).toBeInTheDocument()
    })

    it('renders skeleton loading text when data is null and loading is true', async () => {
        useData.mockImplementation(() => {
            return { data: null, error: false, loading: true }
        })

        act(() => {
            render(<ProductsFilterSection />)
        })

        expect(screen.getByTestId('skeleton-loading-text')).toBeInTheDocument()
    })

    it('renders correctly when data is available and loading is false', async () => {
        useData.mockImplementation(() => {
            return { data: mockCategories, error: false, loading: false }
        })

        act(() => {
            render(<ProductsFilterSection />)
        })

        expect(
            screen.getByTestId('products-filter-section'),
        ).toBeInTheDocument()
        expect(screen.getByTestId('products-filter-section')).toMatchSnapshot()
    })
    
    it('changing category calls handleCategoryChange', async () => {
        const user = userEvent.setup()
        const handleCategoryChangeMock = vi.fn()
        useData.mockImplementation(() => {
            return { data: mockCategories, error: false, loading: false }
        })

        act(() => {
            render(<ProductsFilterSection handleCategoryChange={handleCategoryChangeMock}/>)
        })

        await user.selectOptions(screen.getByLabelText(/category:/i), 'jewelery')

        expect(handleCategoryChangeMock).toHaveBeenCalled()
    })

    it('changing sort calls handleSort', async () => {
        const user = userEvent.setup()
        const handleSortMock = vi.fn()
        useData.mockImplementation(() => {
            return { data: mockCategories, error: false, loading: false }
        })

        act(() => {
            render(<ProductsFilterSection handleSort={handleSortMock}/>)
        })

        await user.selectOptions(screen.getByLabelText(/sort:/i), 'A-Z')

        expect(handleSortMock).toHaveBeenCalled()
    })

    it('searching for a product calls handleSearchInputChange', async () => {
        const user = userEvent.setup()
        const handleSearchInputChangeMock = vi.fn()
        useData.mockImplementation(() => {
            return { data: mockCategories, error: false, loading: false }
        })

        act(() => {
            render(<ProductsFilterSection handleSearchInputChange={handleSearchInputChangeMock}/>)
        })

        await user.type(screen.getByLabelText(/search:/i), 't-shirt')

        expect(handleSearchInputChangeMock).toHaveBeenCalled()
    })
})
