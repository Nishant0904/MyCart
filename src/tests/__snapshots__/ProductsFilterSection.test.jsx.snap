// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Products Filter Section test > renders correctly when data is available and loading is false 1`] = `
<section
  class="sc-guDLey RUbxG"
  data-testid="products-filter-section"
>
  <h1 />
  <div
    class="sc-dmyCSP bYuVhk"
  >
    <div>
      <label
        for="search-input"
      >
        Search:
      </label>
      <input
        autocomplete="off"
        id="search-input"
        type="text"
      />
    </div>
    <div>
      <label
        for="category-select"
      >
        Category:
      </label>
      <select
        id="category-select"
      >
        <option
          value="all"
        >
          All
        </option>
        <option
          value="electronics"
        >
          Electronics
        </option>
        <option
          value="jewelery"
        >
          Jewelery
        </option>
        <option
          value="men's clothing"
        >
          Men's clothing
        </option>
        <option
          value="women's clothing"
        >
          Women's clothing
        </option>
      </select>
    </div>
    <div>
      <label
        for="sort-select"
      >
        Sort:
      </label>
      <select
        id="sort-select"
      >
        <option
          value="id-ascending"
        >
          Default (by SKU)
        </option>
        <option
          value="title-ascending"
        >
          A-Z
        </option>
        <option
          value="title-descending"
        >
          Z-A
        </option>
        <option
          value="price-ascending"
        >
          Price (low to high)
        </option>
        <option
          value="price-descending"
        >
          Price (high to low)
        </option>
        <option
          value="rating-descending"
        >
          Rating (high to low)
        </option>
      </select>
    </div>
  </div>
</section>
`;
